src/modules/adaptors/Balancer/BalancerPoolAdaptor.sol

```diff
@@ -0,0 +1,295 @@
+
+    /**
+     * @notice call `BalancerRelayer` on mainnet to carry out txs that can incl. joins, exits, and swaps
+     * @param tokensIn specific tokens being input for tx
+     * @param amountsIn amount of assets input for tx
+     * @param bptOut acceptable amount of assets resulting from tx (due to slippage, etc.)
+     * @param callData encoded specific txs to be used in `relayer.multicall()`
+     * NOTE: The difference btw what Balancer outlines to do and what we want to do is that we want the txs to be carried out by the vault. Whereas we could just end up connecting to the relayer and have it carry out multi-txs which would be bad cause we need to confirm each step of the multi-call.
+     * TODO: rename params and possibly include others for `exit()` function. ex.) bptOut needs to be renamed to be agnostic.
+     * TODO: see issue for strategist callData
+     */
+    function useRelayer(
+        ERC20[] memory tokensIn,
+        uint256[] memory amountsIn,
+        ERC20 bptOut,
+        bytes[] memory callData
+    ) public {
+        for (uint256 i; i < tokensIn.length; ++i) tokensIn[i].approve(relayer(), amountsIn[i]); // TODO: this may be obsolete based on approval steps necessary for relayer functionality (see TODO below)
+
+        uint256 startingBpt = BPTout.balanceOf(address(this));
+
+        // TODO: implement proper way of giving Relayer approval; Relayer.setRelayerApproval() or Relayer.approveVault().
+
+        relayer().multicall(callData);
+
+        uint256 endingBpt = BPTout.balanceOf(address(this));
+
+        uint256 amountBptOut = endingBpt - startingBpt;
+
+        uint256 amountBptIn = priceRouter.getValues(tokensIn, amountsIn, bptOut); // TODO: not 100% sure how this math all works out. I think this roughly calcs the BPTValue In, it depends on the source of its priceData cause it has to know how much slippage to account for.
+
+        // check value in vs value out
+        if (amountBptOut < amountBptIn.mulDivDown(0.9e4, 1e4)) revert("Slippage");
+
+        // revoke token in approval
+        for (uint256 i; i < tokensIn.length; ++i) _revokeExternalApproval(tokensIn[i], relayer());
+
+        // TODO: see if special revocation is required or necessary with bespoke Relayer approval sequences
+    }
+
+    /**
+     * @notice deposit (stake) BPT into respective pool gauge
+     * @param _poolGauge address of rewards gauge
+     * @param _amountIn number of bpts to deposit
+     * @param _claim_rewards whether or not to claim pending rewards too (true == claim)
+     * @dev 1.) Interface custom as Balancer/Curve do not provide for liquidityGauges.
+     * @dev 2.) Strategists to provide _poolGauge (which they can get via off-chain querying methods)
+     * TODO: Finalize interface details when beginning to do unit testing
+     */
+    function depositBPT(address _poolGauge, uint256 _amountIn, bool _claim_rewards) external {
```

Comment:
For all the strategist functions that use `_poolGuage` can they be changed so that they accept a bpt, then use that to query the pool guage address from the pool guage factory?

Reply:
I will look into that, that sounds like an easier UX for strategists I assume?

Reply:
`depositBPT()`, `withdrawBPT()`, and `claimRewards()` are changed so they have an input `param` as the `address _bpt` now.

For `depositBPT()` and `withdrawBPT()` it should make sense! `claimRewards()` is up in the air as outlined in our other conversation in this PR.